File: backend/agent/tools/memory.py
from __future__ import annotations

import asyncio
from typing import Any, Dict

from agent.memory.store import MemoryStore
from agent.tools.base import Tool


class MemoryTool(Tool):
	name = "memory"
	description = (
		"Read or write persistent memory facts that survive across sessions. "
		"action: get (one key or all), set (key + value), delete (key), list (keys), clear (all). "
		"Use to remember user facts, preferences, and project details between sessions."
	)
	params = {
		"action": {"type": "string", "description": "get | set | delete | list | clear"},
		"key": {"type": "string", "description": "Note key (required for set/delete, optional for get)"},
		"value": {"type": "string", "description": "Note value (required for set)"},
	}

	def __init__(self) -> None:
		self._store = MemoryStore()

	def tool_definition(self) -> dict:
		d = super().tool_definition()
		d["function"]["parameters"]["required"] = ["action"]
		return d

	async def execute(self, args: Dict[str, Any]) -> str:
		action = str(args.get("action", "")).strip().lower()
		key = str(args.get("key", "")).strip()
		value = args.get("value")
		try:
			return await asyncio.to_thread(self._run, action, key, value)
		except Exception as e:
			return f"error: {e}"

	def _run(self, action: str, key: str, value: Any) -> str:
		if action == "list":
			entries = self._store.get_entries()
			if not entries:
				return "(empty)"
			return "\n".join(
				sorted({f"{e['category']}/{e['key']}: {e['value']}" for e in entries})
			)

		if action == "get":
			entries = self._store.get_entries()
			if not key:
				if not entries:
					return "(empty)"
				return "\n".join(
					sorted(f"{e['category']}/{e['key']}: {e['value']}" for e in entries)
				)
			matches = [e for e in entries if key.lower() in e["key"].lower()]
			if not matches:
				return f"(no note for '{key}')"
			return "\n".join(
				f"{e['category']}/{e['key']}: {e['value']}" for e in matches
			)

		if action == "set":
			if not key:
				return "error: key is required for set"
			if value is None:
				return "error: value is required for set"
			val = value if isinstance(value, str) else str(value)
			self._store.upsert_entry(
				session_id="tool",
				category="memory",
				key=key,
				value=val,
				confidence=1.0,
			)
			return f"saved: {key}"

		if action == "delete":
			if not key:
				return "error: key is required for delete"
			deleted = self._store.delete_entry(key)
			if deleted == 0:
				return f"(no note for '{key}')"
			return f"deleted: {key}"

		if action == "clear":
			count = self._store.clear_entries()
			return f"cleared {count} notes"

		return "error: action must be get, set, delete, list, or clear"

File: agent/tools/memory.go
package tools

import (
	"fmt"
	"strings"

	"github.com/elev1e1nSure/warden/agent/memory"
)

func openMemory() *memory.MemoryStore {
	s := memory.NewMemoryStore("")
	_ = s.Init()
	return s
}

// --- MemoryListTool ---

type MemoryListTool struct{}

func (t *MemoryListTool) Name() string { return "memory_list" }

func (t *MemoryListTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "List stored memory entries, optionally filtered by category.",
		Params: map[string]any{
			"category": prop("string", "Filter by category (e.g. preference, user, project). Omit to list all."),
		},
	}
}

func (t *MemoryListTool) Execute(args map[string]any) Result {
	category := getStr(args, "category")
	store := openMemory()
	defer store.Close()

	entries, err := store.GetEntries("", category)
	if err != nil {
		return R("error: " + err.Error())
	}
	if len(entries) == 0 {
		return R("no memory entries found")
	}

	var sb strings.Builder
	for _, e := range entries {
		fmt.Fprintf(&sb, "[%s] %s = %s (confidence: %.2f)\n", e.Category, e.Key, e.Value, e.Confidence)
	}
	return R(strings.TrimRight(sb.String(), "\n"))
}

// --- MemorySaveTool ---

type MemorySaveTool struct{}

func (t *MemorySaveTool) Name() string { return "memory_save" }

func (t *MemorySaveTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Save or update a memory entry by category and key.",
		Params: map[string]any{
			"category": prop("string", "Category (e.g. preference, user, project)"),
			"key":      prop("string", "Unique key within the category"),
			"value":    prop("string", "Value to store"),
		},
		Required: []string{"category", "key", "value"},
	}
}

func (t *MemorySaveTool) Execute(args map[string]any) Result {
	category := strings.TrimSpace(getStr(args, "category"))
	key := strings.TrimSpace(getStr(args, "key"))
	value := strings.TrimSpace(getStr(args, "value"))

	if category == "" || key == "" || value == "" {
		return R("error: category, key, and value are required")
	}

	store := openMemory()
	defer store.Close()

	if err := store.UpsertEntry("", category, key, value, 1.0, ""); err != nil {
		return R("error: " + err.Error())
	}
	return R(fmt.Sprintf("saved: [%s] %s = %s", category, key, value))
}

// --- MemoryDeleteTool ---

type MemoryDeleteTool struct{}

func (t *MemoryDeleteTool) Name() string { return "memory_delete" }

func (t *MemoryDeleteTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Delete a memory entry by key.",
		Params: map[string]any{
			"key": prop("string", "Key of the entry to delete"),
		},
		Required: []string{"key"},
	}
}

func (t *MemoryDeleteTool) Execute(args map[string]any) Result {
	key := strings.TrimSpace(getStr(args, "key"))
	if key == "" {
		return R("error: key is required")
	}

	store := openMemory()
	defer store.Close()

	n, err := store.DeleteEntry(key)
	if err != nil {
		return R("error: " + err.Error())
	}
	if n == 0 {
		return R(fmt.Sprintf("no entry found with key '%s'", key))
	}
	return R(fmt.Sprintf("deleted %d entry(s) with key '%s'", n, key))
}

// --- MemoryClearTool ---

type MemoryClearTool struct{}

func (t *MemoryClearTool) Name() string { return "memory_clear" }

func (t *MemoryClearTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Clear all memory entries.",
		Params:      map[string]any{},
	}
}

func (t *MemoryClearTool) Execute(args map[string]any) Result {
	store := openMemory()
	defer store.Close()

	n, err := store.ClearEntries("")
	if err != nil {
		return R("error: " + err.Error())
	}
	return R(fmt.Sprintf("cleared %d memory entries", n))
}
